File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        map<Node*,Node*> mp;
        Node* temp=head;
        while(temp){
            mp[temp]= new Node(temp->val);
            temp= temp->next;
        }
        temp=head;
        Node* res= new Node(-1), *ans=res;
        while(temp!=NULL){
            res->next=mp[temp];
            res=res->next;
            res->next= mp[temp->next];
            res->random= mp[temp->random];
            res=res->next;
            temp= temp->next;
        }
        return ans->next;
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head==null) return null;
        HashMap<Node, Node> adds= new HashMap<>();
        Node ans=new Node(head.val), itr= ans;
        adds.put(head, ans);
        for(Node temp= head.next; temp!=null;  temp= temp.next){
            itr.next= new Node(temp.val);
            itr= itr.next;
            adds.put(temp, itr);
        }
        itr= ans;
        for(Node temp= head; temp!=null; temp= temp.next){
            if(temp.random!=null) itr.random= adds.get(temp.random);
            itr= itr.next;
        }
        return ans;
    }
}